rework handling of ICMP error messages

yomimono opened the pull request:
Previously, we discarded most ICMP error messages because the
encapsulated IPv4 packet appeared to be incorrectly truncated, as we
were using the `tcpip` library's IPv4 header unmarshaller,
which checked the length field of IPv4 packets against the length
of the payload, and would return an error if they did not match.  This
isn't appropriate for the ICMP error case, where the message is
truncated.  `tcpip` now exposes a function for just parsing the header,
which we use here to get the header information and correctly process
the error packet.

There are also some style changes which clarify handling for this nested
case where ICMP error messages encapsulate an IPv4 packet, and
comments on how this case should be handled in a NAT device and why.

Co-authored-by: linse <linse@linse.me>

yomimono commented:
@xaki23 , with this set of patches to `mirage-nat` we're able to get useful output from `tracepath` in a downstream appvm, and also to receive encapsulated ICMP error messages of the type you mentioned in [your comment on qubes-mirage-firewall's related issue](https://github.com/mirage/qubes-mirage-firewall/issues/31#issuecomment-480123400):

```
[user@untrusted-af ~]$ ping -s 1000 -Q 0xbe 1.2.3.4
PING 1.2.3.4 (1.2.3.4) 1000(1028) bytes of data.
From 10.137.0.5 icmp_seq=1 Destination Port Unreachable
From 10.137.0.5 icmp_seq=2 Destination Port Unreachable
^C
```

Let us know whether this works for you!

yomimono commented:
Test builds using the `mirage-xen` package are currently failing in the sandboxed context, but the interested user can verify that the build works locally.

xaki23 commented:
attempted with the following two pins:

`opam pin -y add tcpip https://github.com/mirage/mirage-tcpip.git`
`opam pin -y add mirage-nat https://github.com/mirage/mirage-nat.git#separate-header`

if i dont pin tcpip, i get a 3.7.3, with the pin its the expected 3.7.4.
with the 3.7.3 i got to the point where it somehow blew up in mirage-xen land.
with the (pinned-to-master) 3.7.4 i get deps problems building mirage-nat:

> mirage-nat is now pinned to git+https://github.com/mirage/mirage-nat.git#separate-header (version 1.1.0)
>
> The following dependencies couldn't be met:
>  - mirage-nat → lru < 0.3.0
>  - mirage-nat → tcpip >= 3.7.2 → lru >= 0.3.0
> Your request can't be satisfied:
>  - No available version of lru satisfies the constraints

xaki23 commented:
if you want to try the qubes-ish build ... take a look at the cut-n-paste ready comments at the beginning of https://github.com/QubesOS/qubes-builder/blob/master/example-configs/mirage.conf

if you want to avoid ssh-agent build, remove all mentions of ssh-agent from builder.conf after copying it from examples.

to add the pins, after the "make get-sources", add them to qubes-src/mirage-firewall/Makefile.builder like so ... https://github.com/mirage/qubes-mirage-firewall/blob/04bea6e9baf4f449252dd7d9730ab54301c70e14/Makefile.builder
(and branches are anchors, and module name doesnt have to match git repo name, but i guess how to deal with opam may be more obvious to you than to me...)

as usual, feedback and suggestions welcome.

xaki23 commented:
i am also moderately sure the nat is still doing bad things in some cases.

in particular, when NATing in an icmp error it seems to modify the _source_ of the icmp to the _dest_ of the original pkt/session, even if the icmp was really generated by some completely different ip (some router/fw in between).

because the icmp "this wont work" _can_ be sent by some-random-node-in-between we have this "include the first chunk of the paket that triggered this icmp". 

once i get this to build again i will post detailed comparison of mirage-fw vs linux-fw for these usecases. 
this is just a heads-up because i am quite sure to have seen it do this, but i couldnt tell you which code version that was.

yomimono commented:
>  with the 3.7.3 i got to the point where it somehow blew up in mirage-xen land.

Details?

 > with the (pinned-to-master) 3.7.4 i get deps problems building mirage-nat:

There's a PR for mirage-nat to build with the version of `lru` that `tcpip` version 3.7.4 expects over at #28 .  I've pushed this branch, rebased on that branch, to `separate-header-new-tcpip` - you can pin it with `opam pin add mirage-nat https://github.com/mirage/mirage-nat.git#separate-header-new-tcpip`.

> in particular, when NATing in an icmp error it seems to modify the source of the icmp to the dest of the original pkt/session, even if the icmp was really generated by some completely different ip (some router/fw in between).

I also observed this and added [some tests for it](https://github.com/mirage/mirage-nat/blob/separate-header-new-tcpip/lib_test/test_rewrite.ml#L338) and [code to fix it](https://github.com/mirage/mirage-nat/blob/separate-header-new-tcpip/lib/nat_rewrite.ml#L232).

xaki23 commented:
now with these two pins ... 
`opam pin -y add tcpip https://github.com/mirage/mirage-tcpip.git`
`opam pin -y add mirage-nat https://github.com/mirage/mirage-nat.git#separate-header-new-tcpip`

it blows up in mirage-xen land again:
```
[ERROR] The compilation of mirage-xen failed at "/home/user/.opam/opam-init/hooks/sandbox.sh build env OPAM_PKG_CONFIG_PATH=/home/user/.opam/mirage-firewall/lib/pkgconfig dune build -p mirage-xen -j 1".

#=== ERROR while compiling mirage-xen.3.3.0 ===================================#
# context              2.0.3 | linux/x86_64 | ocaml-base-compiler.4.07.1 | https://opam.ocaml.org#5b9162fe
# path                 ~/.opam/mirage-firewall/.opam-switch/build/mirage-xen.3.3.0
# command              ~/.opam/opam-init/hooks/sandbox.sh build env OPAM_PKG_CONFIG_PATH=/home/user/.opam/mirage-firewall/lib/pkgconfig dune build -p mirage-xen -j 1
# exit-code            1
# env-file             ~/.opam/log/mirage-xen-23679-580388.env
# output-file          ~/.opam/log/mirage-xen-23679-580388.out
### output ###
# Path outside the workspace: ../pkgconfig/mirage-xen.pc from .
```
i am aware of ...

> failing in the sandboxed context, but the interested user can verify that the build works locally.

... but this interested user doesnt know how to work around this.

yomimono commented:
Sorry for that comment - I don't have this failure for sandbox escape locally and was under the impression that it was a CI artifact that wouldn't be reproduced for someone who tried it interactively.  I'll see whether I can reproduce this and figure out where it's coming from.  Can you let me know what version of `opam` you're running (`opam --version`)?

xaki23 commented:
nothing to be sorry about there, my comment wasnt a complaint. i was just pointing out (again) that i am not much of an ocaml person, so i might need hints for things i cant figure out by "randomly pressing buttons until i find a combo that works". 

did some random button pressing just now (after reading the PR by hannesm).
just nuking state and trying again didnt work (even though the PR says merged).
adding another pin (dune to 1.9.0) makes the build work, the template installs, starts, and passes basic functional testing including ICMP.

will do and post the promised more detailed comparison for the icmp cases in the next days.

i recommend to get the currently-pinned changes merged/pushed/published already, then we can do the next round of icmp detail/edgecase improvements (if needed) against this new baseline instead of causing more confusion.

and by currently-pinned i mean:
```
opam pin -y add dune 1.9.0
opam pin -y add tcpip https://github.com/mirage/mirage-tcpip.git
opam pin -y add mirage-nat https://github.com/mirage/mirage-nat.git#separate-header-new-tcpip
```

hannesm commented:
Dear @xaki23, from the three pins you have in your list, two shouldn't be needed anymore (if you've an up-to-date opam repository (`opam update` should do the trick -- that one syncs from `opam.ocaml.org` via https, and is thus a few hours behind the github.com/ocaml/opam-repository)):
- dune 1.9.0 (since mirage-xen now conflicts with dune 1.9.1 -- see https://github.com/ocaml/opam-repository/pull/13944 https://github.com/ocaml/opam-repository/pull/13945)
- tcpip (since 3.7.4 was released, that plays well with lru 0.3.0 -- https://github.com/ocaml/opam-repository/pull/13929)

> i recommend to get the currently-pinned changes merged/pushed/published already, then we can do the next round of icmp detail/edgecase improvements (if needed) against this new baseline instead of causing more confusion.

I agree, and as mentioned above, only the mirage-nat pin should be necessary.

xaki23 commented:
> if you've an up-to-date opam repository (`opam update` should do the trick

this turns out not to be the case.
when i wrote "nuking state" earlier, i meant "rm -rf ~/.opam ; opam init -y" (which i hope would be  similar to "opam update" in terms of result).
does opam keep state/cache outside ~/.opam somewhere?

just to make sure it isnt something that changed in the last few hours, i removed the dune-1.9.0 pin, tried both "nuke" and "opam update" approaches, and got dune 1.9.1 and the "path outside the workspace" error again. 

>  tcpip (since 3.7.4 was released, that plays well with lru 0.3.0 -- [ocaml/opam-repository#13929 (https://github.com/ocaml/opam-repository/pull/13929))

this is indeed resolved without pinning now.

so right now i am down to these two pins for a working qubes-builder build of mirage-firewall:
```
opam pin -y add dune 1.9.0
opam pin -y add mirage-nat https://github.com/mirage/mirage-nat.git#separate-header-new-tcpip
```

(i only tested build, not install+run on this just now)

xaki23 commented:
some more digging ... the current (2019-04-15 07-ish) delay of opam repo data seems to be "several days". (or perhaps "until someone pushes the right button")

doing a "wget -m https://opam.ocaml.org/index.tar.gz" results in a file stamped 2019-04-12 16:19 GMT, this matches with timestamps inside that tarball (f.ex. the "repo" file).

comparing to https://github.com/ocaml/opam-repository/pull/13944#issuecomment-482638835  ... mouseover the "X days ago" ... says 2019-04-12 16:29 GMT.

actualy, quick bisect on the data contained in the index against the commit history for /opam-repository/ says the last change included is ... https://github.com/ocaml/opam-repository/commit/5b9162fe5d249c1f5659e916d924cd291c555fbd
so about 4 days // 35 commits are missing currently. 
"ups".

yomimono commented:
Since this looks good in tests which are working and has been user-validated, I'll merge.

xaki23 commented:
i can confirm the opam index is updated now and the dune pin is no longer required.
i assume this merge here means i can remove the #branch from the mirage-nat pin, but still need to pin it to the git version for now since mirage-nat has not been rolled into a new version yet?

yomimono commented:
> i assume this merge here means i can remove the #branch from the mirage-nat pin, but still need to pin it to the git version for now since mirage-nat has not been rolled into a new version yet?

Assuming that "rolled into a new version" means a new version of qubes-mirage-firewall, yes, that's correct.  If you mean that `mirage-nat` hasn't yet been released in a new version incorporating these changes, that's also pending (and therefore currently demands a pin), but hopefully will be merged soon as `mirage-nat` version 1.2.0 -- see https://github.com/ocaml/opam-repository/pull/13973 for current status.
